\documentclass[a4paper,12pt]{article}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{booktabs}
\usepackage{natbib}
\usepackage{amsmath}
\usepackage[colorlinks=true]{hyperref}
\usepackage[left=2cm]{geometry}


\title{Testing the unguarded X hypothesis}
\begin{document}
\maketitle
\section{Introduction}
The unguarded X (UX) hypothesis states that one reason for females to live longer than males is that they are usually the homogametic sex. This implies that either somatic or inherited (partially) recessive deleterious mutations in the X chromosome are always exposed in males, but concealed in heterozygous females. This hypothesis requires recessive mutations, maintained by mutation-selection balance, that negatively affect longevity both in males and in females. This scenario is hardly controversial. The main uncertainty is whether this kind of mutations can fully explain the difference in longevity between sexes.

The alternative hypothesis is that a different kind of mutations is required to account for the difference in longevity between males and females. Namely, sexually antagonistic mutations, which are maintained at relatively high frequencies in the population by a selection balance. Sexually antagonistic mutations could happen in any chromosome. However, there are theoretical reasons and evidence to belive that they are especially frequent in the X chromosome \citep{Gibson2002}.

\citet{Kelly1999} suggested an experiment to determine if low-frequency, deleterious mutations can explain the observed genetic variance in a character that is correlated with fitness. The alternative is that high-frequency mutations maintained by a selection balance are required to explain the observed variance. The experiment is relatively easy and it involves a short selection experiment and a short inbreeding experiment. If most genetic variance in the character is due to a mutation-selection balance, the response to selection, small as it may be, will be comparable to the effect of inbreeding. However, in the case of any selection balance (antagonistic pleiotropy or overdominance), the response to selection is expected to be very low, while the effect of inbreeding relatively high. Needless to say that the conclusion would only apply to the population used in the experiment.

Kelly probably did not have in mind a sex-specific character when he suggested such an experiment. And he did not attempt to guess on what chromosome the relevant variants were located. Thus, we need to make sure if the UX hypothesis can be tested with a similar experiment, and exactly how it should be performed. The right thing to do would be to derive the expectations of the relevant quantities (covariance of additive effects and homozygous dominance effects, $C_{ad}$, and additive variance, $V_a$) of a dimorphic character taking into account that at least some loci must be in the X chromosome. This may have been done, but I have not found it yet in the literature.

Before getting there, I want to note that \citet{Charlesworth1996} assume frequencies of mutant alleles are low, even though they may affect survival only late in life. This is also an assumption we need to justify the application of Kelly's method to test the UX hypothesis. It could be questioned, because a low selection coefficient against such mutations suggests they could behave as (almost) neutral and eventually drift to not-so-low frequencies.

\section{Genetic variance components}
\citet{Charlesworth1996} cite \citet{Falconer1989} and \citet{Mukai1974} as sources of the standard formula for additive genetic variance. In chapter 7, \citet{Falconer1989} assigns genotypic values $-a$, $d$, and $a$ to genotypes A$_2$A$_2$, A$_1$A$_2$, and A$_1$A$_1$, respectively. If the frequency of allele A$_1$ is $p$ and that of A$_2$ is $q$, and assuming Hardy-Weinberg, the average genotypic value at one locus is $a(p-q)+2dpq$ \citep[p. 114]{Falconer1989}.

\citet{Mukai1974} considers a locus affecting fitness, and therefore uses a different parameterization. The genotypic values (fitness) corresponding to genotypes A$_2$A$_2$, A$_1$A$_2$, and A$_1$A$_1$ (again in increasing order) are now $1-s$, $1-hs$, and 1, respectively. In this case, assuming Hardy-Weinberg equilibrium, the average genotypic value is $1-2pqhs-q^2s$. Note that in this case we could also assume that the population is at mutation-selection equilibrium. Then, according to classic theory, the frequency of A$_2$ should be something like $\mu/hs$.

The total genetic variance due to one locus with two alleles can be easily derived for a population in Hardy-Weinberg equilibrium by finding the expected squared deviation from the mean of the genotypic values. Total genetic variance at one locus can be decomposed in additive, $V_A$, and dominance, $V_D$, components. Table~\ref{tau:variance} shows de expressions of genetic variance components found in both sources. \citet[p. 135]{Falconer1989} gives detais of how to derive those expressions.

\begin{table}
\begin{center}
\caption{Genetic variance components at one autosomal locus, expressed according to either \citet{Falconer1989} or \citet{Mukai1974}.}\label{tau:variance}
\vspace*{0.3cm}
\begin{tabular}{lll}
\toprule
Component&Falconer&Mukai et al.\\
\midrule
Additive&$2pq[a+d(q-p)]^2$&$2pqs^2[(p-q)h+q]^2$\\
Dominance&$(2pqd)^2$&$p^2q^2s^2(1-2h)^2$\\
Total&$2pq[a + d(q-p)]^2 + [2pqd]^2$&$pqs^2[2(1-2pq)h^2 - 4q^{2}h + q(1+q)]$\\
\bottomrule
\end{tabular}
\end{center}
\end{table}

So far, the locus is assumed to be autosomal. What would the genetic variance of an X-linked locus be?

\section{Genetic variance of an X-linked locus}
I assume for simplicity that sex-ratio is balanced, and that the heterogametic (male) genotypes have the same genotypic values as the corresponding homozygotes (see Table~\ref{tau:xlinked}). The average genotypic value of an X-linked gene is $a(p-q)+pqd$, which is slightly different from the mean genotypic value of an autosomal locus ($a(p-q)+2pqd$). The average gene effects also change. If my calculations are correct, the average effect of allele A$_1$ is $\alpha{}_1 = q(\frac{3}{2}a + d(\frac{1}{2}-p))$, and that of allele A$_2$ is $\alpha{}_2 = -p(\frac{3}{2}a + d(q-\frac{1}{2}))$. The average effect of substituting an A$_2$ allele for an A$_1$ is $\alpha = \frac{3}{2}a + \frac{1}{2}d(q-p)$. Note that the following relationships hold for both autosomal and X-linked genes: $\alpha = \alpha{}_1 - \alpha{}_2$; $\alpha{}_1 = q\alpha$; and $\alpha{}_2 = -p\alpha$.

The breeding value of a genotype is twice the expected genotypic value of its progeny, expressed usually as deviation from the population average.

\begin{table}
\begin{center}
\caption{}\label{tau:xlinked}
\vspace*{0.3cm}
{\small
\begin{tabular}{p{2.4cm}cccccc}
\toprule
&A$_1$A$_1$&A$_1$A$_2$&A$_2$A$_2$&A$_1$0&A$_2$0&Mean\\
\midrule
Frequency&$\frac{1}{2}q^2$&$pq$&$\frac{1}{2}q^2$&$\frac{1}{2}p$&$\frac{1}{2}q$&\\
Gen. value&$a$&$d$&$-a$&$a$&$-a$&$a(p-q)+pqd$\\
Deviance&$q(2a-pd)$&$a(q-p)+d(1-pq)$&$-p(2a+qd)$&$q(2a-pd)$&$-p(2a+qd)$&0\\
Breed. val.&&&&&&\\
\bottomrule
\end{tabular}
}
\end{center}
\end{table}

\section{Apendix}

\bibliographystyle{abbrvnat}
\bibliography{ux}

\end{document}
